File: src/java/magma/option/Option.java
package magma.option;

import magma.collect.list.List_;
import magma.compile.Node;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public interface Option<T> {
    <R> Option<R> map(Function<T, R> mapper);

    T orElseGet(Supplier<T> other);

    Tuple<Boolean, T> toTuple(T other);

    void ifPresent(Consumer<T> consumer);

    T orElse(T other);

    Option<T> filter(Predicate<T> predicate);

    boolean isPresent();

    <R> R match(Function<T, R> ifPresent, Supplier<R> ifEmpty);

    boolean isEmpty();

    Option<T> or(Supplier<Option<T>> other);

    <R> Option<R> flatMap(Function<T, Option<R>> mapper);
}

File: src/clang/magma/option/Option.c
#include "Option.h"
magma.option.Option<magma.option.R> map(magma.option.R(*mapper)(T));
T orElseGet(T(*other)());
magma.option.Tuple<int, T> toTuple(T other);
magma.option.void ifPresent(void(*consumer)(T));
T orElse(T other);
magma.option.Option<T> filter(int(*predicate)(T));
int isPresent();
magma.option.R match(magma.option.R(*ifPresent)(T), magma.option.R(*ifEmpty)());
int isEmpty();
magma.option.Option<T> or(magma.option.Option<T>(*other)());
magma.option.Option<magma.option.R> flatMap(magma.option.Option<magma.option.R>(*mapper)(T));

